### PR TITLE
ADD: UPGRADE for ETH.RUNE

### LINKED ISSUE #1326: ADD: UPGRADE for ETH.RUNE
Add UPGRADE tag and function for ETH.RUNE

`0x3155ba85d5f96b2d030a4966af206230e46849cb`

It is simply `SWITCH:thoraddress` via the Router

### PR DESCRIPTION
**Pull request by thatStrangeGuyThorchain**
- added `upgrade` button for all possible `NonNativeRune` assets for `AssetsTableCollapsable` and `AssetDetails`
- refactored `UpgradeView` and `upgrade` service to handle all available `NonNativeRune` assets
- splitted `UpgradeView` to separate ones to have more control on fees (now only fees)
- added loading decimals info for `UpgradeView`-related-stuff
- changed `LoadingView`

closes #1326

### CODE REVIEW

File: src/renderer/components/wallet/assets/AssetDetails.tsx [hunk 1]
```diff
@@ -70,11 +70,11 @@ export const AssetDetails: React.FC<Props> = (props): JSX.Element => {
     )
   }, [oWalletAddress, history])
 
-  const walletActionUpgradeRuneBnbClick = useCallback(() => {
+  const walletActionUpgradeNonNativeRuneClick = useCallback(() => {
     FP.pipe(
       oWalletAddress,
-      O.filter((_) => AssetHelper.isRuneBnbAsset(asset)),
-      O.map((walletAddress) => walletRoutes.upgradeBnbRune.path({ asset: assetToString(asset), walletAddress })),
+      O.filter((_) => AssetHelper.isNonNativeRuneAsset(asset)),
```
Comment (veado) [inline src/renderer/components/wallet/assets/AssetDetails.tsx]: [nitpick] As same as `(_) => isNonNativeRuneAsset`

File: src/renderer/components/wallet/assets/AssetDetails.tsx [hunk 2]
```diff
@@ -99,42 +99,46 @@ export const AssetDetails: React.FC<Props> = (props): JSX.Element => {
...
 
   const isRuneNativeAsset: boolean = useMemo(() => eqAsset.equals(asset, AssetRuneNative), [asset])
 
-  const oRuneBnbBalance: O.Option<WalletBalance> = useMemo(
-    () => getWalletBalanceByAsset(oBalances, oRuneBnbAsset),
-    [oRuneBnbAsset, oBalances]
-  )
-
-  const oRuneBnbAmount: O.Option<BaseAmount> = useMemo(
+  const getNonNativeRuneBalance = useMemo(
```
Comment (veado) [inline src/renderer/components/wallet/assets/AssetDetails.tsx]: [nitpick] Add type signature ^
`const getNonNativeRuneBalance: O.Option<(balances: WalletBalances) => O.Option<AssetAmount>>`

File: src/renderer/components/wallet/assets/AssetDetails.tsx [hunk 3]
```diff
@@ -99,42 +99,46 @@ export const AssetDetails: React.FC<Props> = (props): JSX.Element => {
...
-  const runeUpgradeDisabled: boolean = useMemo(
-    () =>
-      isRuneBnbAsset &&
+  const runeUpgradeDisabled: boolean = useMemo(() => {
+    return (
+      isNonNativeRuneAsset &&
       FP.pipe(
-        oRuneBnbAmount,
+        oNonNativeRuneAmount,
         O.map((amount) => amount.amount().isLessThan(0)),
```
Comment (veado) [inline src/renderer/components/wallet/assets/AssetDetails.tsx]: [nitpick] `amount.amount().isLessThan(0)` -> `amount.lt(0)`

File: src/renderer/components/wallet/txs/upgrade/Upgrade.tsx [hunk 4]
```diff
@@ -100,25 +98,54 @@ export const Upgrade: React.FC<Props> = (props): JSX.Element => {
...
     setAmountToUpgrade(ZERO_BASE_AMOUNT)
-  }, [resetUpgradeTxState])
+    /**
+     * As `Upgrade` component depends on parent's Component state we also have to trigger reload of parent component's state too
+     *
+     * I.E. Loading of targetPoolAddressRD was failed:
+     * Upgrade component renders Error view AND HAS TO:
+     *    1. reset its' own state
+     *    2. trigger higher-level data update
+     */
```
Comment (veado) [inline src/renderer/components/wallet/txs/upgrade/Upgrade.tsx:108]: Such comments are always appreciated!

File: src/renderer/views/wallet/upgrade/UpgradeView.tsx [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        return <UpgradeETH {...props} />
+      }
+      default:
+        return null
+    }
+  }, [])
+
+  const reloadOnError = useCallback(() => {
+    // Reload inbound addresses in case previously it was failed
+    FP.pipe(O.some(reloadInboundAddresses), O.ap(O.fromPredicate(RD.isFailure)(targetPoolAddressRD)))
```
Comment (veado) [inline src/renderer/views/wallet/upgrade/UpgradeView.tsx]: Suggestion to keep it more (easily) readable:
```
    FP.pipe(
      targetPoolAddressRD,
      O.fromPredicate(RD.isFailure),
      O.map((_) => reloadInboundAddresses)
    )
```

File: src/renderer/helpers/assetHelper.ts [hunk 6]
```diff
@@ -54,6 +56,13 @@ export const getBnbRuneAsset = (network: Network = 'testnet'): Asset => {
 export const isRuneBnbAsset = (asset: Asset): boolean =>
   eqAsset.equals(asset, AssetRune67C) || eqAsset.equals(asset, AssetRuneB1A)
 
+export const isRuneEthAsset = (asset: Asset): boolean =>
+  eqAsset.equals(asset, AssetRuneERC20) ||
+  eqAsset.equals(asset, AssetRuneERC20Testnet) ||
+  eqAsset.equals(asset, AssetRuneEthERC20)
+
+export const isNonNativeRuneAsset = (asset: Asset): boolean => isRuneBnbAsset(asset) || isRuneEthAsset(asset)
```
Comment (veado) [inline src/renderer/helpers/assetHelper.ts:64]: Why not `!isRuneNativeAsset(asset)` ?

**Review by veado — CHANGES_REQUESTED** [review]
Nice work! _Just few nitpick ..._

File: src/renderer/components/wallet/txs/upgrade/Upgrade.tsx (hunk 4 above)
Comment (thatStrangeGuyThorchain) [reply src/renderer/components/wallet/txs/upgrade/Upgrade.tsx:108]: yeah. i've spent some time to understand this bug which we missed previously...

File: src/renderer/views/wallet/upgrade/UpgradeView.tsx (hunk 5 above)
Comment (thatStrangeGuyThorchain) [reply src/renderer/views/wallet/upgrade/UpgradeView.tsx]: np. but:
```ts
FP.pipe(
      targetPoolAddressRD,
      O.fromPredicate(RD.isFailure),
      O.map(reloadInboundAddresses) // direct call of reloadInboundAddresses needed
    )
```
P.S.  fully functional notation (with `O.ap`) does not allow to forget such things 😅

File: src/renderer/helpers/assetHelper.ts (hunk 6 above)
Comment (thatStrangeGuyThorchain) [reply src/renderer/helpers/assetHelper.ts:64]: suppose here is some misunderstanding...
```ts
!isRuneNativeAsset(asset)
```
is about to check if asset is not `THOR.RUNE`. it means `!isRuneNativeAsset(BNB.BNB) => true` which is not we want here. 
with `isNonNativeRuneAsset` we want to get all `[CHAIN].RUNE` assets except `THOR.RUNE`
```ts
isNonNativeRuneAsset(BNB.RUNE) === true
isNonNativeRuneAsset(ETH.RUNE) === true
// ^these cases are the only ones now who returns true results^
isNonNativeRuneAsset(THOR.RUNE) === false
isNonNativeRuneAsset(BNB.BNB) === false
isNonNativeRuneAsset(BCH.BCH) === false
// ... etc
```

Comment (veado) [reply src/renderer/helpers/assetHelper.ts:64]: You are right! _I've missed that ... ^_

**Review by veado — APPROVED** [review]
:+1:

**Pull request merged**